PostgreSQL interval datatype support

Envek opened the pull request:
Add native support for PostgreSQL `interval` datatype with conversion from all PostgreSQL formats to `ActiveSupport::Duration` on record load and transformation from `ActiveSupport::Duration` to ISO 8601 on save (no more strings). Also added support to define column in migrations and get it in schema dump, optional column precision is supported.
- Parses ISO 8601 duration format, tells Postgres to use it (see [Output formats](http://www.postgresql.org/docs/9.4/static/datatype-datetime.html#DATATYPE-INTERVAL-OUTPUT))
- Constructs `ActiveSupport::Duration` value (that magic `2.days`)
- Converts `ActiveSupport::Duration` to [ISO 8601 duration](http://en.wikipedia.org/wiki/ISO_8601#Durations) format on save
- Do not allows to use old raw values on save — you must ensure you're feeding correct ISO strings to model object.
- Allows to use `Numeric` values, treating them as number of seconds (as some calculations with Time objects returns not durations but floats)

Example:

```
create_table :events do |t|
  t.string   :name
  t.interval :duration
end

Event.create!(name: 'Rock Fest', duration: 2.days)
Event.last.duration # => 2 days
Event.last.duration.iso8601 # => "P2D"
```

I've proposed this in Rails-core Google Group here: https://groups.google.com/forum/#!topic/rubyonrails-core/zcscu1p4ucA

Envek commented:
Implemented `type_cast_from_user` method in [activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb](https://github.com/Envek/rails/blob/pg_interval_support/activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb#L20)

voltechs commented:
@Envek which gist did you update? Can't seem to get the 4.2 initializer (you posted earlier) working...

sgrif commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -461,7 +462,6 @@ def initialize_type_map(m) # :nodoc:
           m.register_type 'ltree', OID::SpecializedString.new(:ltree)
 
           # FIXME: why are we keeping these types as strings?
-          m.alias_type 'interval', 'varchar'
```
Unfortunately this is a breaking change, so we can't actually do this without a deprecation cycle. We should continue to treat it as a string and emit a warning, since w/ the attributes API you can now specify manually in the mean time.

Envek replied:
Can you point me on some example how to do that (deprecation warning and some kind of switch to enable this feature immediately)?

sgrif replied:
Basically, do this:

```ruby
m.register_type 'interval' do |*args, sql_type|
  unless ActiveRecord::Base.real_interval_types
    ActiveSupport::Deprecation.warn(<<-MSG.squish)
      blah blah blah deprecated
    MSG
    m.lookup(varchar, *args)
  else
    # what you have now
  end
end
```

Envek replied:
@sgrif Pushed deprecation warning in bc366e1, is it good for you?

Envek commented:
@voltechs I have tested initializer today in 4.2.4 app, it's working. Here it's current version: https://gist.githubusercontent.com/Envek/7077bfc36b17233f60ad/raw/25d12443c0fdf0dccf0f51275688c1e0264c55ec/pg_interval_support_4_2.rb

If you still can't get it working, please contact me directly, as this conversation is for pull request related staff only. Thanks.

sgrif commented:
@Envek are you still interested in working on this? It'll need some more work to be mergeable into master.

Envek commented:
@sgrif sure I still interested in all my PRs. Will add deprecation warning tomorrow.

Please note that this PR is based on top of #16917, so it should be merged first (I have split ActiveSupport and ActiveRecord stuff on separate PRs). Could you please take a look on it? I will do my best to finish both these PRs. Let's merge them! :smile:

ollym commented:
This is really important for me too, have a load of hacks in our code base that's keeping this together.

ollym commented:
@Envek I highly recommend you check out the excellent work @jeremyevans is has done with the PG interval column within sequel here:
https://github.com/jeremyevans/sequel/blob/master/lib/sequel/extensions/pg_interval.rb

It even contains a full PostgreSQL Interval parser.

sgrif commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb:
```diff
@@ -0,0 +1,42 @@
+          end
+
+          def type_cast_from_database(value)
+            if value.kind_of? ::String
```
Do we expect values other than a string or `nil` here? I'd rather not let those slip through. What do you think of changing to this instead?

```ruby
def type_cast_from_database(value)
  return if value.nil?
  ::ActiveSupport::Duration.parse(value)
end
```

sgrif commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb:
```diff
@@ -0,0 +1,42 @@
+          def type_cast_from_user(value)
+            type_cast_from_database(value)
+          rescue ::ActiveSupport::Duration::ISO8601ParsingError
+            value # Allow user to supply raw string values in another formats supported by PostgreSQL
```
I think we should restrict this only to formats which we can parse for the time being. I don't like the idea of this type returning a string some of the time. We can always add support for more formats later.

Envek replied:
Restricted. It will raise `ActiveSupport::Duration::ISO8601Parser::ParsingError` exception and will say what value is wrong:

    > Event.new(duration: '1 year 2 mons')
    ActiveSupport::Duration::ISO8601Parser::ParsingError: Invalid ISO 8601 duration: "1 year 2 mons"

sgrif commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb:
```diff
@@ -0,0 +1,42 @@
+            case value
+              when ::ActiveSupport::Duration
+                value.iso8601(self.precision)
+              when ::Numeric
```
Do we actually receive numerics here? Can we drop this branch?

Envek replied:
Unfortunately some math with `Time`s results not in `Duration`s being returned, but `Float`s with number in seconds:
```
Time.current - (Time.current + 1.hour) # => -3600.00000453 (Float)
```
I think that we should be able to handle that.

sgrif commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -494,6 +494,21 @@ def initialize_type_map(m) # :nodoc:
+              Working with PostgreSQL interval datatype as with varchar is deprecated!
+              Starting from 5.1 interval values will be converted to and from ActiveSupport::Duration objects by default.
+              To use new behavior please set ActiveRecord::Base.postgresql_interval_type_as_duration to true in config/application.rb
+              See https://github.com/rails/rails/pull/16919 for details.
```
This will print out a deprecation warning for all apps on boot. We should do this as I suggested in my previous comment so it only affects people who are currently using `interval` columns.

sgrif replied:
Actually, we shouldn't even handle this here. This should be added to https://github.com/rails/rails/blob/9deb6ab/activerecord/lib/active_record/model_schema.rb#L371-L390

sgrif commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -667,6 +682,9 @@ def configure_connection
           end
 
+          # Set interval output format to ISO 8601 for ease of parsing by ActiveSupport::Duration.parse
+          execute('SET intervalstyle = iso_8601', 'SCHEMA')
```
It's a shame that the Ruby PG gem (and libpq) don't allow for receiving certain columns in binary, but others as text. `interval` is significantly easier to deal with in the binary form (64 bit signed integer representing number of microseconds, 32 bit signed integer representing number of days, 32 bit signed integer representing number of months).

Envek commented:
@sgrif #16917 has landed to the master branch and that's a good time to ressurect this pull request!

I've applied your notes, rebased on top of current master (more than 10 000 commits forward!) and changed implementation to match current ActiveRecord.

Please review again.

Envek commented:
@sgrif Do you still want this PR to be included in 5.0? I want it to be on board (sure if we still can catch this train before departure).

Can you please review?

sgrif commented:
No, we've removed it from the 5.0 milestone. It'll be merged shortly after rc1 is released.

Envek commented:
As RC1 is released now, I have changed 5.0 and 5.1 to 5.1 and 5.2 accordingly in deprecation warnings and rebased on top of current master.

Also added a section to Guides page about PostgreSQL support.

sgrif commented:
Thank you. I will be moving through the backlog this week.

On Mon, May 9, 2016, 1:22 PM Andrey Novikov <notifications@github.com>
wrote:

> As RC1 is released now, I have changed 5.0 and 5.1 to 5.1 and 5.2
> accordingly in deprecation warnings and rebased on top of current master.
>
> Also added a section to Guides page about PostgreSQL support.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/16919#issuecomment-217929653>
>

boazy commented:
I've tried to monkey patch ActiveRecord with the changes in this PR, but it looks like the ISO 8601 duration serialization/parsing methods introduced in #16917 produce inconsistent results when months and years are used together. The following tests may fail (success depends on *the current time*, which makes it even worse):

```ruby
ActiveSupport::Duration.parse(2.months.iso8601) == 2.months
ActiveSupport::Duration.parse(3.years.iso8601) == 3.years
```

The inconsistency comes from the fact that months and years (and technically even [days](https://en.wikipedia.org/wiki/Leap_second)) are not fixed durations. Months can range from 28 to 31 days, days range from 86359 to 86401 seconds, and years range from 365 to 366 days.
`ActiveSupport::Duration` wisely ignores all this complexity and just assigns fixed values to each duration part:
```ruby
1.year == 365.25.days
1.month == 30.days
1.day == 86400.seconds
```
This is probably the best approach, since it produces consistent results and keeps durations independent of the current time. `ActiveSupport::Duration` still tracks the date/time parts, so ``Time.parse('2015-02-15') + 1.month`` is 2015-03-15 and ``Time.parse('2016-02-15') + 1.month`` is 2016-03-15 even though February was 28 days long in 2015 and 29 days long in 2016.

To keep in line with this approach, we can change ``ActiveSupport::Duration.parse`` to this code:
```ruby
class ActiveSupport::Duration
  def self.parse(iso8601duration)
    parts = ISO8601Parser.new(iso8601duration).parse!
    total_seconds = 0
    parts.each do |part, value|
      total_seconds += value.send part.to_sym
    end
    new(total_seconds, parts)
  end
end
```

Envek commented:
@boazy Please take a look at issue https://github.com/rails/rails/issues/26941 and fixing it commits https://github.com/rails/rails/commit/b0b4a9a65861ce05042f334a81c70b1365762e98 and https://github.com/rails/rails/commit/8931916f4a1c1d8e70c06063ba63928c5c7eab1e.

Most probably your issue is fixed in not yet released Rails 5.0.1, try to use rails from `5-0-stable` branch. For that, replace `gem 'rails'` in your Gemfile with:

```ruby
gem 'rails', '~> 5.0.0', github: 'rails/rails', branch: '5-0-stable'
```

sgrif commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb:
```diff
@@ -0,0 +1,42 @@
+              value
+            when ::String
+              # Will raise +ActiveSupport::Duration::ISO8601Parser::ParsingError+ on string of unknown format
+              ::ActiveSupport::Duration.parse(value)
```
This method should not raise on an invalid string. It should return `nil` in order to be consistent with the rest of the Active Record types.

Envek replied:
Now it returns `nil`.

boazy commented:
@Envek Yes, I've seen the commit, but I've tested this issue against master, and it's not fixed.
The problem is that using an EPOCH just fixes the DST issue for short periods (based on the assumption there is no DST in January), but it still suffer from the same major fault - you probably shouldn't rely on ``Time#advance`` at all when calculating durations, since the value of `month`, `year` and possibly `second` (I'm not sure how Ruby deals with leap seconds) are not the same as they are for `ActiveSupport::Duration`.

I'll try to submit a pull request fixing this issue with unit tests which show the problem. I already have it part-done.

crashtech commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb:
```diff
@@ -0,0 +1,42 @@
+              # Example: Time.current - (Time.current + 1.hour) # => -3600.000001776 (Float)
+              time = ::Time.current
+              duration = ::ActiveSupport::Duration.new(time.advance(seconds: value) - time, seconds: value)
+              duration.iso8601(precision: self.precision)
```
@Envek Wouldn't be easier to rely on ```.seconds``` instead of creating the dependency with machine's time?

```ruby
when ::Numeric
  value.seconds.iso8601(precision: self.precision)
```

Envek replied:
Done. Thank you.

Envek commented on activerecord/lib/active_record/model_schema.rb:
```diff
@@ -499,6 +500,30 @@ def compute_table_name
+          else
+            array_arguments = ""
+          end
+          if column.sql_type.start_with?("interval")
```
@sgrif, I've noticed that you've removed this deprecation warning for points in [`b347156`](https://github.com/rails/rails/commit/b347156b15aea7f6229203e0cc1a58937ecb8e93#diff-6c8fd220c92c2b25aa29891b59d00c04), but it's not clear to me whether I should to remove it too or not. I decided to keep it for now. Please suggest the proper solution here.

sgrif replied:
Yes, you will need to provide a deprecation warning. There's no reason to have a `legacy_interval` type, the user can just do `attribute :foo, :string`

Envek replied:
Removed `legacy_interval` type, adjusted changelog and deprecation warning accordingly

Envek commented:
As this PR is quite old, I had to rebase it against fresh master to resolve multiple merge conflicts (mostly about replacement of single quotes to double quotes). Meanwhile addressed reviews and also renamed deprecated interval-as-a-string type from `interval_string` to `legacy_interval` to keep naming consistent with implementation of points datatype.

I would be very grateful if this PR will be reviewed one more time.

simi commented on guides/source/active_record_postgresql.md:
```diff
@@ -399,6 +399,29 @@ macbook.address
+* [type definition](http://www.postgresql.org/docs/current/static/datatype-datetime.html#DATATYPE-INTERVAL-INPUT)
+* [functions and operators](http://www.postgresql.org/docs/current/static/functions-datetime.html)
+
+This type is mapped to Ruby on Rails [`ActiveSupport::Duration`](http://api.rubyonrails.org/classes/ActiveSupport/Duration.html) objects.
```
I think you can omit `Ruby on Rails` mention here.

Envek replied:
Done

sgrif commented:
@matthewd Can you take over this one for me? I'm going to be unavailable for a bit, would like to get this into 5.1 though.

kurko commented:
@matthewd @sgrif anything anyone can do to help move this forward?

Envek commented:
Yeah, I'll be glad to resurrect it, rebase and fix any new review comments if anyone will be ready to review it. Didn't follow the development of ActiveRecord last year though.

craineum commented:
Would like to move to `schema.rb` and away from `structure.sql`, but this is currently preventing us. @Envek any movement on this? Thanks for your time.

voltechs commented:
Good lord I want this so bad. Thanks for your work on this, @Envek.

I don't mean to be pessimistic, but... if this hasn't attracted the attention by now that it needs, do we think it will just sitting here as an open PR? I've never opened a PR against a project as big as Rails—what are other ways we can get some eyes and love on this puppy?

crashtech commented:
Hey Guys, I'm not sure if I can do this, but until this is merged, you can use a GEM that I've created. I implemented the interval datatype and many other features. [Torque PostgreSQL](https://rubygems.org/gems/torque-postgresql/versions/0.2.3)

kofronpi commented:
Sad to see this useful PR hanging for so long ! Are there any plans to merge it ?

marnen commented:
I’d like to get interval support into Rails. How can I help get this pull request ready to merge?

marnen commented:
It would still be nice to get this merged (once conflicts are resolved). This is a fairly important outstanding issue, so I don't really want the bots to close it.

rafaelfranca commented:
We will need a volunteer to resolve the conflicts and push the PR to the finish line, otherwise in 3 months the bot will mark the PR as stale again and we'll have to close.

marnen commented:
@rafaelfranca Now that you’ve clarified that, perhaps we stand a better chance of having it happen. :) (I might give it a try, but I’m already overcommitted...)

voltechs commented:
It seems as though we need some dedicated support/eyes from the internal rails core team. I'm sensing rebase fatigue from the contributors here. Feels a little tone-deaf to doom this PR to the dustbin simply because there hasn't been a lot of support from the merge guardians. Is anyone prepared to review/accept this PR of contributors step up? Nobody likes wasting other people's time, let alone their own.

Thanks for stepping up, @marnen. I too am considering helping with this, but I'm also fairly overcommitted. I at least wanted to chime in and refresh my support (for what little it's worth) for this.

marnen commented:
@voltechs I agree. Note that I'm trying *not* to have to step up, first because I don't actually have time, and second because I'm putting a little more effort into https://github.com/crashtech/torque-postgresql/issues/41 , which sort of covers the same ground. 

But I do need this functionality in my Rails applications, and so I will ultimately attempt to deal with this if no one with more time (and more skill hacking ActiveRecord!) comes forward to do it.

Envek committed:
Add support for PostgreSQL `interval` datatype.

Add support for PostgreSQL `interval` data type with conversion to
`ActiveSupport::Duration` when loading records from database and
serialization to ISO 8601 formatted duration string on save.
Add support to define a column in migrations and get it in a schema dump.
Optional column precision is supported.

To use this in 6.1, you need to place the next string to your model file:

    attribute :duration, :interval

To keep old behavior until 6.2 is released:

    attribute :duration, :string

Envek commented:
I have refreshed this pull request. Can someone review it again? @rafaelfranca, or can you maybe recommend someone to ask for a review?

Due to many merge conflicts, I had to squash all commits (that addressed previous reviews) into one, or it was seemed to be impossible to resolve all the conflicts. However, none of the conflicts required significant changes, so in reality, this pull request stayed mostly the same. I even restored some tests that I deleted earlier. And, sure, I bumped version numbers in deprecation warning and changelog.

Thanks everyone for your support comments – they helped me not to forget about this change, and I returned to it as soon as there was a spare minute.

Envek commented on activerecord/lib/active_record/connection_adapters/postgresql/oid/interval.rb:
```diff
@@ -0,0 +1,88 @@
+              if match[:seconds]
+                seconds_parse_method = /\d+\.\d+/ =~ match[:seconds] ? :to_f : :to_i
+                parts[:seconds] = sign * timesign * match[:seconds].send(seconds_parse_method)
+              end
```
It was extracted into https://github.com/rails/rails/pull/16917 and merged a long, long time ago. For some reason this thread wasn't marked as stale.

Envek commented:
Updated gist with initializers for adding support for interval data type into existing applications. Added initializer for both Rails 5.2 and 6.0: https://gist.github.com/Envek/7077bfc36b17233f60ad#file-pg_interval_support_6_0-rb

AgentAntelope commented:
Hey there, me and @h-lame were looking at using this and found that when using the gist example, we needed to add a little more code to support schema dumping for default values. We added:

```
          def type_cast_for_schema(value)
            serialize(value).inspect
          end
```

to `ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Interval` in order to support this. Hope this helps get this in! Would definitely like to see this support in core Rails.

Envek commented:
@AgentAntelope, nice catch, thank you! Added method from you to the datatype along with a test case.

marnen commented:
Ongoing issue, should still be open.

jeremy commented:
Let's get this in. This effort is old enough to be entering grade school. Thank you for your patience & persistence @Envek.

Envek commented:
> Let's get this in.

This is fantastic! :partying_face: Thank you!

I rebased PR on top of current `master` branch, fixed conflicts, checked that Rails version in deprecation warning is up to date. Just in case :wink:

jeremy commented:
Merged @ 0475215d4fa1a6db2a92a0065081fe19c64cc124
